### PR TITLE
Complete record synchronization

### PR DESCRIPTION
**Pull request by bengaal**
Create records with invenio-records-marc21 and insert file attachments (WIP).

### CODE REVIEW

File: invenio_rdm_pure/source/rdm/run/synchronizer.py [hunk 1]
```diff
@@ -12,67 +12,84 @@
...
-        """Default Constructor of the class Synchronizer."""
+    def __init__(
+        self,
+        pure_api_url: str = "",
+        pure_api_key: str = "",
+        pure_username: str = "",
+        pure_password: str = "",
+    ):
+        """Default Constructor of the Synchronizer class."""
+        if not pure_api_url:
```
Comment (utnapischtim) [inline invenio_rdm_pure/source/rdm/run/synchronizer.py:45]: it would be nicer to have the following statements in the lines 40 - 43. so not to have there empty strings but the real default value

File: invenio_rdm_pure/source/rdm/database.py [hunk 2]
```diff
@@ -38,28 +39,33 @@ def select_query(self, fields: str, table: str, filters={}):
...
-            invenio_pure_user = datastore.get_user(invenio_pure_user_email)
-            if not invenio_pure_user:
-                invenio_pure_user = datastore.create_user(
-                    email=invenio_pure_user_email,
-                    password=hash_password(invenio_pure_user_password),
-                    active=True,
+        if not self.pure_user_id:
+            datastore = current_app.extensions["security"].datastore
+            if datastore is not None:
+                invenio_pure_user_email = current_app.config.get(
```
Comment (utnapischtim) [inline invenio_rdm_pure/source/rdm/database.py]: it would be better to move those things also to the parameter list of the method. so to set default parameters in the method. so it is easier to test this method

File: setup.py [hunk 3]
```diff
@@ -42,6 +42,7 @@
     "invenio-app>=1.3.0",
     "invenio-celery>=1.2.2",
     "invenio-db>=1.0.8",
+    "invenio-pidstore>=1.2.2",
```
Comment (utnapischtim) [inline setup.py]: question: why is this necessary?

File: invenio_rdm_pure/source/utils.py [hunk 4]
```diff
@@ -24,6 +24,15 @@
 )
 
 
+def load_file_as_string(path):
```
Comment (utnapischtim) [inline invenio_rdm_pure/source/utils.py:27]: could you please add some error handling in the case the file at the path does not exists

File: invenio_rdm_pure/source/rdm/run/synchronizer.py [hunk 5]
```diff
@@ -81,25 +98,79 @@ def synchronize_research_outputs(
...
+                        self.delete_record_files(files)
             except RuntimeError:
                 traceback.print_exc()
 
-        # TODO: Store record with the help of marc21 module
-        pass
+    def create_record(self, record_xml: str) -> RecordItem:
+        """Create Invenio record from Marc21XML string."""
+        identity = Identity(self.pure_user_id)
+        identity.provides.add(any_user)
```
Comment (utnapischtim) [inline invenio_rdm_pure/source/rdm/run/synchronizer.py:130]: i am not sure if it is wise to add any_user to the identity. but i have also no suggestions, so go ahead with it.

File: invenio_rdm_pure/source/rdm/run/synchronizer.py (hunk 1 above)
Comment (bengaal) [reply invenio_rdm_pure/source/rdm/run/synchronizer.py:45]: I don't think it would work and would be good practice. Default method arguments should be compile-time constants.

Comment (utnapischtim) [reply invenio_rdm_pure/source/rdm/run/synchronizer.py:45]: a function call is possible. I didn't kew compile-time constants exists in python. But it is ok, let it as it is

File: setup.py (hunk 3 above)
Comment (bengaal) [reply setup.py]: The `RecordItem` object the `Marc21RecordServices` returns only has the `Record` object as a private member, accessing it would be possible but bad practice. Additionally the aforementioned member somehow isn't identical with the `Record` object I would expect it to be. In order to access the proper record I would use the `Resolver` from the invenio-pidstore module, which would return the proper `Record`, so files can be attached.

File: invenio_rdm_pure/source/rdm/database.py (hunk 2 above)
Comment (bengaal) [reply invenio_rdm_pure/source/rdm/database.py]: What do you have in mind specifically?

File: invenio_rdm_pure/source/utils.py (hunk 4 above)
Comment (bengaal) [reply invenio_rdm_pure/source/utils.py:27]: Error handling added.

File: invenio_rdm_pure/source/rdm/run/synchronizer.py (hunk 1 above)
Comment (bengaal) [reply invenio_rdm_pure/source/rdm/run/synchronizer.py:45]: Otherwise it throws a `RuntimeError: Working outside of application context.` error.

Comment (utnapischtim) [reply invenio_rdm_pure/source/rdm/run/synchronizer.py:45]: ahhhh, because of the current_app part. then let it as it is.

File: invenio_rdm_pure/source/rdm/database.py (hunk 2 above)
Comment (utnapischtim) [reply invenio_rdm_pure/source/rdm/database.py]: i had in mind to move the current_app.config calls to the parameter default sets, but this is not possible, see you message above. 
i prefer to have pure functions, so to give those values as parameters to the function

Comment (bengaal) [reply invenio_rdm_pure/source/rdm/database.py]: I have modified the function, I hope it is what you had in mind.

File: setup.py (hunk 3 above)
Comment (bengaal) [reply setup.py]: I have deleted the dependency and the reference.

**Review by utnapischtim — APPROVED** [review]
